### DISCUSSION TITLE
Mobx react: shared observable renders twice?

### DISCUSSION THREAD

**Discussion by rickmed** [mobxjs/mobx#3535] [Q&A] [answered]
If an observable is used in more than one place in the ui with react, since (as I understand correctly), mobx-react will call setState on an effect, will react render twice? (even though conceptually should be one atomic ui change)

**Comment by kubk** [comment]
Why do you think it's going to render twice? By the way you can always check re-renders by yourself. Either using React devtools extension or by simply adding console.log to a component

**Comment by urugator** [comment, marked as answer]
> will react render twice?

No, individual setState/forceUpdate calls are wrapped in single [`unstable_batchedUpdates`](https://github.com/mobxjs/mobx/blob/4f8b014116024965c5edab40539e5eb6ed771a6a/packages/mobx-react-lite/src/index.ts#L9) (no longer neccessary in react >= 18). React then handles cross-component batching - setState/forceUpdate only marks component as dirty and at the end of `unstable_batchedUpdates` (or microtask in react >18), the component tree is traversed from root and dirty components are re-rendered.
[`react-dom.flushSync`](https://reactjs.org/docs/react-dom.html#flushsync) can be used to re-render immediately.

**Comment by mweststrate** [comment]
It will render in one atomic update
